File: source/lib/samba3/upgrade.c
/* 
   Unix SMB/CIFS implementation.
   Generate ldb_message 's for samba3_*

    Copyright (C) Jelmer Vernooij 	2005
   
   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.
   
   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.
   
   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/

#include "includes.h"
#include "lib/samba3/samba3.h"
#include "lib/ldb/include/ldb.h"

static struct ldb_message *msg_array_add(struct ldb_context *ctx, struct ldb_message ***msgs, int *count)
{
	struct ldb_message *ret;
	*msgs = talloc_realloc(ctx, *msgs, struct ldb_message *, (*count)+1);

	ret = (*msgs)[*count] = talloc_zero(ctx, struct ldb_message);
	(*count)++;

	return ret;
}

static struct ldb_dn *regkey_to_dn(struct ldb_context *ldb, const char *name)
{
	char *p, *n, *dup;
	struct ldb_dn *ret = ldb_dn_explode(ldb, "hive=NONE");

	p = dup = talloc_strdup(ldb, name);

	while (p) {
		n = strchr(p, '/');
		if (n) { *n = '\0';	n++; }

		ret = ldb_dn_build_child(ldb, "key", p, ret);

		p = n;
	}

	talloc_free(dup);

	return ret;
}

/* Where prefix is any of:
 * - HKLM
 *   HKU
 *   HKCR
 *   HKPD
 *   HKPT
 */

int samba3_upgrade_registry(struct samba3_regdb *regdb, const char *prefix, struct ldb_context *ldb, struct ldb_message ***msgs)
{
	int i;
	struct ldb_message *msg;
	int count = 0;
	char *prefix_up = strupper_talloc(ldb, prefix);
	*msgs = NULL;

	for (i = 0; i < regdb->key_count; i++) {
		int j;
		struct samba3_regkey *rk = &regdb->keys[i];
		struct ldb_dn *keydn;

		/* Only handle selected hive */
		if (strncmp(prefix_up, rk->name, strlen(prefix_up)) != 0) {
			continue;
		}

		msg = msg_array_add(ldb, msgs, &count);

		msg->num_elements = 0;
		msg->elements = NULL;
		msg->private_data = NULL;

		/* Convert key name to dn */
		keydn = msg->dn = regkey_to_dn(ldb, rk->name);

		ldb_msg_add_string(ldb, msg, "name", strrchr(rk->name, '/')?strrchr(rk->name, '/')+1:rk->name);
		
		for (j = 0; j < rk->value_count; j++) {
			struct samba3_regval *rv = &rk->values[j];

			msg = msg_array_add(ldb, msgs, &count);
			msg->dn = ldb_dn_build_child(ldb, "value", rv->name, keydn);

			ldb_msg_add_string(ldb, msg, "value", rv->name);
			ldb_msg_add_fmt(ldb, msg, "type", "%d", rv->type);
			ldb_msg_add_value(ldb, msg, "data", &rv->data);
		}
	}
	
	talloc_free(prefix_up);

	return count;
}

int samba3_upgrade_sam(struct samba3 *samba3, struct ldb_context *ldb, struct ldb_message ***msgs)
{
	int count = 0;
	struct ldb_message *msg;
	struct ldb_dn *domaindn = NULL;
	const char *domainname;
	struct samba3_domainsecrets *domsec;
	int i;
	*msgs = NULL;

	domainname = samba3_get_param(samba3, "global", "workgroup");

	if (domainname == NULL) {
		DEBUG(0, ("No domain name specified in smb.conf!\n"));
		return -1;
	}

	domsec = samba3_find_domainsecrets(samba3, domainname);

	/* Domain */	
	msg = msg_array_add(ldb, msgs, &count);
	/* FIXME: Guess domain DN by taking ldap bind dn? */

	ldb_msg_add_string(ldb, msg, "objectClass", "top");
	ldb_msg_add_string(ldb, msg, "objectClass", "domain");
	ldb_msg_add_string(ldb, msg, "objectSid", dom_sid_string(msg, &domsec->sid));
	ldb_msg_add_string(ldb, msg, "objectGUID", GUID_string(msg, &domsec->guid));
	ldb_msg_add_string(ldb, msg, "name", domainname);
	ldb_msg_add_string(ldb, msg, "oEMInformation", "Provisioned by Samba4 (upgraded from Samba3)");

	/* account policy as well */

	ldb_msg_add_fmt(ldb, msg, "minPwdLength", "%d", samba3->policy.min_password_length);
	ldb_msg_add_fmt(ldb, msg, "pwdHistoryLength", "%d", samba3->policy.password_history);
	ldb_msg_add_fmt(ldb, msg, "minPwdAge", "%d", samba3->policy.minimum_password_age);
	ldb_msg_add_fmt(ldb, msg, "maxPwdAge", "%d", samba3->policy.maximum_password_age);
	ldb_msg_add_fmt(ldb, msg, "lockoutDuration", "%d", samba3->policy.lockout_duration);
	ldb_msg_add_fmt(ldb, msg, "samba3ResetCountMinutes", "%d", samba3->policy.reset_count_minutes);
	ldb_msg_add_fmt(ldb, msg, "samba3UserMustLogonToChangePassword", "%d", samba3->policy.user_must_logon_to_change_password);
	ldb_msg_add_fmt(ldb, msg, "samba3BadLockoutMinutes", "%d", samba3->policy.bad_lockout_minutes);
	ldb_msg_add_fmt(ldb, msg, "samba3DisconnectTime", "%d", samba3->policy.disconnect_time);
	ldb_msg_add_fmt(ldb, msg, "samba3RefuseMachinePwdChange", "%d", samba3->policy.refuse_machine_password_change);
	
	/* Users */
	for (i = 0; i < samba3->samaccount_count; i++) {
		struct samba3_samaccount *sam = &samba3->samaccounts[i];

		msg = msg_array_add(ldb, msgs, &count);
		msg->dn = ldb_dn_build_child(msg, "cn", sam->fullname, domaindn);

		ldb_msg_add_string(ldb, msg, "objectClass", "top");
		ldb_msg_add_string(ldb, msg, "objectClass", "person");
		ldb_msg_add_string(ldb, msg, "objectClass", "user");
		ldb_msg_add_fmt(ldb, msg, "lastLogon", "%d", sam->logon_time);
		ldb_msg_add_fmt(ldb, msg, "lastLogoff", "%d", sam->logoff_time);
		ldb_msg_add_string(ldb, msg, "unixName", sam->username);
		ldb_msg_add_string(ldb, msg, "name", sam->nt_username);
		ldb_msg_add_string(ldb, msg, "cn", sam->fullname);
		ldb_msg_add_string(ldb, msg, "description", sam->acct_desc);
		ldb_msg_add_fmt(ldb, msg, "primaryGroupID", "%d", sam->group_rid); 
		ldb_msg_add_fmt(ldb, msg, "badPwdcount", "%d", sam->bad_password_count);
		ldb_msg_add_fmt(ldb, msg, "logonCount", "%d", sam->logon_count);
		
		ldb_msg_add_string(ldb, msg, "samba3Domain", sam->domain);
		if (sam->dir_drive) 
			ldb_msg_add_string(ldb, msg, "samba3DirDrive", sam->dir_drive);

		if (sam->munged_dial)
			ldb_msg_add_string(ldb, msg, "samba3MungedDial", sam->munged_dial);

		if (sam->homedir)
			ldb_msg_add_string(ldb, msg, "samba3Homedir", sam->homedir);

		if (sam->logon_script)
			ldb_msg_add_string(ldb, msg, "samba3LogonScript", sam->logon_script);

		if (sam->profile_path)
			ldb_msg_add_string(ldb, msg, "samba3ProfilePath", sam->profile_path);

		if (sam->workstations)
			ldb_msg_add_string(ldb, msg, "samba3Workstations", sam->workstations);

		ldb_msg_add_fmt(ldb, msg, "samba3KickOffTime", "%d", sam->kickoff_time);
		ldb_msg_add_fmt(ldb, msg, "samba3BadPwdTime", "%d", sam->bad_password_time);
		ldb_msg_add_fmt(ldb, msg, "samba3PassLastSetTime", "%d", sam->pass_last_set_time);
		ldb_msg_add_fmt(ldb, msg, "samba3PassCanChangeTime", "%d", sam->pass_can_change_time);
		ldb_msg_add_fmt(ldb, msg, "samba3PassMustChangeTime", "%d", sam->pass_must_change_time);
		ldb_msg_add_fmt(ldb, msg, "samba3Rid", "%d", sam->user_rid); 
	
		/* FIXME: Passwords */
	}

	/* Groups */
	for (i = 0; i < samba3->group.groupmap_count; i++) {
		struct samba3_groupmapping *grp = &samba3->group.groupmappings[i];

		msg = msg_array_add(ldb, msgs, &count);

		if (grp->nt_name != NULL) 
			msg->dn = ldb_dn_build_child(msg, "cn", grp->nt_name, domaindn);
		else 
			msg->dn = ldb_dn_build_child(msg, "cn", dom_sid_string(msg, grp->sid), domaindn);

		ldb_msg_add_string(ldb, msg, "objectClass", "top");
		ldb_msg_add_string(ldb, msg, "objectClass", "group");
		ldb_msg_add_string(ldb, msg, "description", grp->comment);
		ldb_msg_add_string(ldb, msg, "cn", grp->nt_name);
		ldb_msg_add_string(ldb, msg, "objectSid", dom_sid_string(msg, grp->sid));
		ldb_msg_add_string(ldb, msg, "unixName", "FIXME");
		ldb_msg_add_fmt(ldb, msg, "samba3SidNameUse", "%d", grp->sid_name_use);
	}

	return count;
}

int samba3_upgrade_winbind(struct samba3 *samba3, struct ldb_context *ldb, struct ldb_message ***msgs)
{
	int i;
	int count = 0;
	struct ldb_message *msg;
	struct ldb_dn *basedn = NULL;
	*msgs = NULL;

	msg = msg_array_add(ldb, msgs, &count);

	msg->dn = basedn; 
	
	ldb_msg_add_fmt(ldb, msg, "userHwm", "%d", samba3->idmap.user_hwm);
	ldb_msg_add_fmt(ldb, msg, "groupHwm", "%d", samba3->idmap.group_hwm);

	for (i = 0; i < samba3->idmap.mapping_count; i++) {
		char *sid = dom_sid_string(msg, samba3->idmap.mappings[i].sid);
		msg = msg_array_add(ldb, msgs, &count);
		
		msg->dn = ldb_dn_build_child(ldb, "SID", sid, basedn);
		ldb_msg_add_string(ldb, msg, "SID", sid);
		ldb_msg_add_fmt(ldb, msg, "type", "%d", samba3->idmap.mappings[i].type);
		ldb_msg_add_fmt(ldb, msg, "unixID", "%u", samba3->idmap.mappings[i].unix_id);
	}
	
	return count;
}

int samba3_upgrade_winsdb(struct samba3 *samba3, struct ldb_context *ldb, struct ldb_message ***msgs)
{
	int i;
	int count = 0;
	
	for (i = 0; i < samba3->winsdb_count; i++) {
		struct samba3_winsdb_entry *e = &samba3->winsdb_entries[i];
		int j;
		struct ldb_message *msg = msg_array_add(ldb, msgs, &count);

		msg->dn = ldb_dn_string_compose(ldb, NULL, "type=%d,name=%s", e->type, e->name);

		ldb_msg_add_string(ldb, msg, "name", e->name);
		ldb_msg_add_fmt(ldb, msg, "type", "%d", e->type);
		ldb_msg_add_string(ldb, msg, "objectClass", "wins");
		ldb_msg_add_fmt(ldb, msg, "nbFlags", "%x", e->nb_flags);
		ldb_msg_add_string(ldb, msg, "expires", 
				  ldap_timestring(msg, e->ttl));

		for (j = 0; j < e->ip_count; j++) {
			ldb_msg_add_string(ldb, msg, "address", sys_inet_ntoa(e->ips[j]));
		}
	}

	return count;
}

File: source4/scripting/python/samba/upgrade.py
#!/usr/bin/python
#
#	backend code for upgrading from Samba3
#	Copyright Jelmer Vernooij 2005-2007
#	Released under the GNU GPL v3 or later
#

"""Support code for upgrading from Samba 3 to Samba 4."""

from provision import findnss
import provision
import grp
import pwd
from uuid import uuid4
from param import default_configuration

def regkey_to_dn(name):
    """Convert a registry key to a DN."""
	dn = "hive=NONE"

    for el in name.split("/")[1:]:
        dn = "key=%s," % el + dn

	return dn

# Where prefix is any of:
# - HKLM
#   HKU
#   HKCR
#   HKPD
#   HKPT
#

def upgrade_registry(regdb,prefix,ldb):
    """Migrate registry contents."""
    assert regdb is not None:
	prefix_up = prefix.upper()
	ldif = []

    for rk in regdb.keys:
		pts = rk.name.split("/")

		# Only handle selected hive
        if pts[0].upper() != prefix_up:
			continue

		keydn = regkey_to_dn(rk.name)

		pts = rk.name.split("/")

		# Convert key name to dn
		ldif[rk.name] = """
dn: %s
name: %s

""" % (keydn, pts[0])
		
        for rv in rk.values:
			ldif[rk.name + " (" + rv.name + ")"] = """
dn: %s,value=%s
value: %s
type: %d
data:: %s""" % (keydn, rv.name, rv.name, rv.type, ldb.encode(rv.data))

	return ldif

def upgrade_sam_policy(samba3,dn):
	ldif = """
dn: %s
changetype: modify
replace: minPwdLength
minPwdLength: %d
pwdHistoryLength: %d
minPwdAge: %d
maxPwdAge: %d
lockoutDuration: %d
samba3ResetCountMinutes: %d
samba3UserMustLogonToChangePassword: %d
samba3BadLockoutMinutes: %d
samba3DisconnectTime: %d

""" % (dn, samba3.policy.min_password_length, 
	samba3.policy.password_history, samba3.policy.minimum_password_age,
	samba3.policy.maximum_password_age, samba3.policy.lockout_duration,
	samba3.policy.reset_count_minutes, samba3.policy.user_must_logon_to_change_password,
	samba3.policy.bad_lockout_minutes, samba3.policy.disconnect_time)
	
	return ldif

def upgrade_sam_account(ldb,acc,domaindn,domainsid):
    """Upgrade a SAM account."""
    if acc.nt_username is None or acc.nt_username == "":
		acc.nt_username = acc.username

    if acc.fullname is None:
		acc.fullname = pwd.getpwnam(acc.fullname)[4]

	acc.fullname = acc.fullname.split(",")[0]

    if acc.fullname is None:
		acc.fullname = acc.username
	
	assert acc.fullname is not None
	assert acc.nt_username is not None

	ldif = """dn: cn=%s,%s
objectClass: top
objectClass: user
lastLogon: %d
lastLogoff: %d
unixName: %s
sAMAccountName: %s
cn: %s
description: %s
primaryGroupID: %d
badPwdcount: %d
logonCount: %d
samba3Domain: %s
samba3DirDrive: %s
samba3MungedDial: %s
samba3Homedir: %s
samba3LogonScript: %s
samba3ProfilePath: %s
samba3Workstations: %s
samba3KickOffTime: %d
samba3BadPwdTime: %d
samba3PassLastSetTime: %d
samba3PassCanChangeTime: %d
samba3PassMustChangeTime: %d
objectSid: %s-%d
lmPwdHash:: %s
ntPwdHash:: %s

""" % (ldb.dn_escape(acc.fullname), domaindn, acc.logon_time, acc.logoff_time, acc.username, acc.nt_username, acc.nt_username, 
acc.acct_desc, acc.group_rid, acc.bad_password_count, acc.logon_count,
acc.domain, acc.dir_drive, acc.munged_dial, acc.homedir, acc.logon_script, 
acc.profile_path, acc.workstations, acc.kickoff_time, acc.bad_password_time, 
acc.pass_last_set_time, acc.pass_can_change_time, acc.pass_must_change_time, domainsid, acc.user_rid,
	ldb.encode(acc.lm_pw), ldb.encode(acc.nt_pw))

	return ldif

def upgrade_sam_group(group,domaindn):
    """Upgrade a SAM group."""
	if group.sid_name_use == 5: # Well-known group
		return None

    if group.nt_name in ("Domain Guests", "Domain Users", "Domain Admins"):
		return None
	
    if group.gid == -1:
		gr = grp.getgrnam(grp.nt_name)
    else:
		gr = grp.getgrgid(grp.gid)

    if gr is None:
		group.unixname = "UNKNOWN"
    else:
		group.unixname = gr.gr_name

	assert group.unixname is not None
	
	ldif = """dn: cn=%s,%s
objectClass: top
objectClass: group
description: %s
cn: %s
objectSid: %s
unixName: %s
samba3SidNameUse: %d
""" % (group.nt_name, domaindn, 
group.comment, group.nt_name, group.sid, group.unixname, group.sid_name_use)

	return ldif

def upgrade_winbind(samba3,domaindn):
	ldif = """
		
dn: dc=none
userHwm: %d
groupHwm: %d

""" % (samba3.idmap.user_hwm, samba3.idmap.group_hwm)

    for m in samba3.idmap.mappings:
		ldif += """
dn: SID=%s,%s
SID: %s
type: %d
unixID: %d""" % (m.sid, domaindn, m.sid, m.type, m.unix_id)
	
	return ldif

def upgrade_wins(samba3):
    """Upgrade the WINS database."""
	ldif = ""
	version_id = 0

    for e in samba3.winsentries:
		now = sys.nttime()
		ttl = sys.unix2nttime(e.ttl)

		version_id+=1

        numIPs = len(e.ips)

        if e.type == 0x1C:
			rType = 0x2
        elif e.type & 0x80:
            if numIPs > 1:
				rType = 0x2
            else:
				rType = 0x1
        else:
            if numIPs > 1:
				rType = 0x3
            else:
				rType = 0x0

        if ttl > now:
			rState = 0x0 # active
        else:
			rState = 0x1 # released

		nType = ((e.nb_flags & 0x60)>>5)

		ldif += """
dn: name=%s,type=0x%02X
type: 0x%02X
name: %s
objectClass: winsRecord
recordType: %u
recordState: %u
nodeType: %u
isStatic: 0
expireTime: %s
versionID: %llu
""" % (e.name, e.type, e.type, e.name, 
   rType, rState, nType, 
   ldaptime(ttl), version_id)

        for ip in e.ips:
			ldif += "address: %s\n" % ip

	ldif += """
dn: CN=VERSION
objectClass: winsMaxVersion
maxVersion: %llu
""" % version_id

	return ldif

def upgrade_provision(lp, samba3):
	subobj = Object()

	domainname = samba3.configuration.get("workgroup")
	
    if domainname is None:
		domainname = samba3.secrets.domains[0].name
		print "No domain specified in smb.conf file, assuming '%s'\n" % domainname
	
	domsec = samba3.find_domainsecrets(domainname)
	hostsec = samba3.find_domainsecrets(hostname())
	realm = samba3.configuration.get("realm")

    if realm is None:
		realm = domainname
		print "No realm specified in smb.conf file, assuming '%s'\n" % realm
	random_init(local)

	subobj.realm        = realm
	subobj.domain       = domainname
	subobj.hostname     = hostname()

	assert subobj.realm is not None
	assert subobj.domain is not None
	assert subobj.hostname is not None

	subobj.HOSTIP       = hostip()
    if domsec is not None:
		subobj.DOMAINGUID   = domsec.guid
		subobj.DOMAINSID    = domsec.sid
    else:
		print "Can't find domain secrets for '%s'; using random SID and GUID\n" % domainname
		subobj.DOMAINGUID = uuid4()
		subobj.DOMAINSID = randsid()
	
    if hostsec:
		subobj.HOSTGUID     = hostsec.guid
    else:
		subobj.HOSTGUID = uuid4()
	subobj.invocationid = uuid4()
	subobj.krbtgtpass   = randpass(12)
	subobj.machinepass  = randpass(12)
	subobj.adminpass    = randpass(12)
	subobj.datestring   = datestring()
	subobj.root         = findnss(pwd.getpwnam, "root")[4]
	subobj.nobody       = findnss(pwd.getpwnam, "nobody")[4]
	subobj.nogroup      = findnss(grp.getgrnam, "nogroup", "nobody")[2]
	subobj.wheel        = findnss(grp.getgrnam, "wheel", "root")[2]
	subobj.users        = findnss(grp.getgrnam, "users", "guest", "other")[2]
	subobj.dnsdomain    = subobj.realm.lower()
	subobj.dnsname      = "%s.%s" % (subobj.hostname.lower(), subobj.dnsdomain)
	subobj.basedn       = "DC=" + ",DC=".join(subobj.realm.split("."))
	rdn_list = subobj.dnsdomain.split(".")
	subobj.domaindn     = "DC=" + ",DC=".join(rdn_list)
	subobj.domaindn_ldb = "users.ldb"
	subobj.rootdn       = subobj.domaindn

	modules_list        = ["rootdse",
					"kludge_acl",
					"paged_results",
					"server_sort",
					"extended_dn",
					"asq",
					"samldb",
					"password_hash",
					"operational",
					"objectclass",
					"rdn_name",
					"show_deleted",
					"partition"]
	subobj.modules_list = ",".join(modules_list)

	return subobj

smbconf_keep = [
	"dos charset", 
	"unix charset",
	"display charset",
	"comment",
	"path",
	"directory",
	"workgroup",
	"realm",
	"netbios name",
	"netbios aliases",
	"netbios scope",
	"server string",
	"interfaces",
	"bind interfaces only",
	"security",
	"auth methods",
	"encrypt passwords",
	"null passwords",
	"obey pam restrictions",
	"password server",
	"smb passwd file",
	"private dir",
	"passwd chat",
	"password level",
	"lanman auth",
	"ntlm auth",
	"client NTLMv2 auth",
	"client lanman auth",
	"client plaintext auth",
	"read only",
	"hosts allow",
	"hosts deny",
	"log level",
	"debuglevel",
	"log file",
	"smb ports",
	"large readwrite",
	"max protocol",
	"min protocol",
	"unicode",
	"read raw",
	"write raw",
	"disable netbios",
	"nt status support",
	"announce version",
	"announce as",
	"max mux",
	"max xmit",
	"name resolve order",
	"max wins ttl",
	"min wins ttl",
	"time server",
	"unix extensions",
	"use spnego",
	"server signing",
	"client signing",
	"max connections",
	"paranoid server security",
	"socket options",
	"strict sync",
	"max print jobs",
	"printable",
	"print ok",
	"printer name",
	"printer",
	"map system",
	"map hidden",
	"map archive",
	"preferred master",
	"prefered master",
	"local master",
	"browseable",
	"browsable",
	"wins server",
	"wins support",
	"csc policy",
	"strict locking",
	"preload",
	"auto services",
	"lock dir",
	"lock directory",
	"pid directory",
	"socket address",
	"copy",
	"include",
	"available",
	"volume",
	"fstype",
	"panic action",
	"msdfs root",
	"host msdfs",
	"winbind separator"]

#
#   Remove configuration variables not present in Samba4
#	oldconf: Old configuration structure
#	mark: Whether removed configuration variables should be 
#		kept in the new configuration as "samba3:<name>"
def upgrade_smbconf(oldconf,mark):
	data = oldconf.data()
	newconf = param_init()

	for (s in data) {
		for (p in data[s]) {
			keep = False
			for (k in smbconf_keep) { 
                if smbconf_keep[k] == p:
					keep = True
					break
			}

            if keep:
				newconf.set(s, p, oldconf.get(s, p))
            elif mark:
				newconf.set(s, "samba3:"+p, oldconf.get(s,p))
		}
	}

    if oldconf.get("domain logons") == "True":
		newconf.set("server role", "domain controller")
    else:
        if oldconf.get("security") == "user":
			newconf.set("server role", "standalone")
        else:
			newconf.set("server role", "member server")

	return newconf

def upgrade(subobj, samba3, message, paths, session_info, credentials):
	ret = 0
	lp = loadparm_init()
	samdb = Ldb(paths.samdb, session_info=session_info, credentials=credentials)

	message("Writing configuration")
	newconf = upgrade_smbconf(samba3.configuration,True)
	newconf.save(paths.smbconf)

	message("Importing account policies")
	ldif = upgrade_sam_policy(samba3,subobj.BASEDN)
	samdb.modify(ldif)
	regdb = Ldb(paths.hklm)

	regdb.modify("
dn: value=RefusePasswordChange,key=Parameters,key=Netlogon,key=Services,key=CurrentControlSet,key=System,HIVE=NONE
replace: type
type: 4
replace: data
data: %d
" % samba3.policy.refuse_machine_password_change)

	message("Importing users")
    for account in samba3.samaccounts:
		msg = "... " + account.username
		ldif = upgrade_sam_account(samdb, accounts,subobj.BASEDN,subobj.DOMAINSID)
        try:
            samdb.add(ldif)
        except LdbError, e:
            # FIXME: Ignore 'Record exists' errors
			msg += "... error: " + str(e)
			ret += 1; 
		message(msg)

	message("Importing groups")
    for mapping in samba3.groupmappings:
		msg = "... " + mapping.nt_name
		ldif = upgrade_sam_group(mapping, subobj.BASEDN)
        if ldif is not None:
            try:
			    samdb.add(ldif)
            except LdbError, e:
                # FIXME: Ignore 'Record exists' errors
				msg += "... error: " + str(e)
				ret += 1
		message(msg)

	message("Importing registry data")
    for hive in ["hkcr","hkcu","hklm","hkpd","hku","hkpt"]:
		message("... " + hive)
		regdb = Ldb(paths[hive])
		ldif = upgrade_registry(samba3.registry, hive, regdb)
		for (var j in ldif) {
			var msg = "... ... " + j
            try:
                regdb.add(ldif[j])
            except LdbError, e:
                # FIXME: Ignore 'Record exists' errors
				msg += "... error: " + str(e)
				ret += 1
			message(msg)

	message("Importing WINS data")
	winsdb = Ldb(paths.winsdb)
	ldb_erase(winsdb)

	ldif = upgrade_wins(samba3)
	winsdb.add(ldif)

	# figure out ldapurl, if applicable
	ldapurl = None
	pdb = samba3.configuration.get_list("passdb backend")
    if pdb is not None:
        for backend in pdb:
            if len(backend) >= 7 and backend[0:7] == "ldapsam":
                ldapurl = backend[7:]

	# URL was not specified in passdb backend but ldap /is/ used
    if ldapurl == "":
	    ldapurl = "ldap://%s" % samba3.configuration.get("ldap server")

	# Enable samba3sam module if original passdb backend was ldap
    if ldapurl is not None:
		message("Enabling Samba3 LDAP mappings for SAM database")

		samdb.modify("""
dn: @MODULES
changetype: modify
replace: @LIST
@LIST: samldb,operational,objectguid,rdn_name,samba3sam
""")

		samdb.add("""
dn: @MAP=samba3sam
@MAP_URL: %s""", ldapurl))

	return ret

def upgrade_verify(subobj, samba3, paths, message):
	message("Verifying account policies")

	samldb = Ldb(paths.samdb)

    for account in samba3.samaccounts:
		msg = samldb.search("(&(sAMAccountName=" + account.nt_username + ")(objectclass=user))")
		assert(len(msg) >= 1)
	
	# FIXME
